src/client/finAuthFetch.ts

```diff
@@ -0,0 +1,144 @@
+import { finCsrfFetch } from './finCsrfFetch'
+
+/**
+ * An enhanced fetch API for a pluggable authentication mechanism
+ *
+ * Transparently handles pre-authentication, authentication fault detection, authentication, and
+ * will replay the requested resource upon successful authentication.
+ *
+ * This fetch function will utilize the finCsrfFetch function internally to execute the request
+ *
+ * Example
+ * ```typescript
+ *
+ * const result = finAuthFetch(request, {
+ * 		authenticator: {
+ * 			isAuthenticated: (response: Response) => response.status !== 401,
+ *			preauthenticate: async (request: RequestInfo) => new Request(request, {headers: {auth_header: '12345'}}),
+ *			authenticate: (response: Response) => return execute_authentication(username, password),
+ *			maxTries: 3
+ * 	}
+ * })
+ * ```
+ *
+ * @link https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/fetch
+ *
+ * @param resource The resource to request.
+ * @param options Optional object containing custom settings and the authenticator.
+ * @returns A promise that resolves to a response object.
+ * @throws Exception on failed authentication
+ */
+export async function finAuthFetch(
+	resource: RequestInfo,
+	options?: RequestInit | RequestInitAuth
+): Promise<Response> {
+	let resp: Response | undefined
+
+	if (options && isAuthInit(options)) {
+		const authenticator = options.authenticator as RequestAuthenticator
+
+		// Pre authenticate if suppprted by authenticator
+		if (authenticator.preauthenticate) {
+			resource = await authenticator.preauthenticate(resource)
+		}
+
+		// Pipe request to the csrf fetch function
+		resp = await finCsrfFetch(resource, options)
+
+		// Check if response is an authentication fault
+		if (!(await authenticator.isAuthenticated?.(resp))) {
+			const maxTries = authenticator.maxTries ?? 3
+			let authCount = 0
+			let authSuccessful = false
+
+			// Authentication generator to generate an authentication call
+			// while we are under the max try threshold
+			const tryAuthentication = async function* () {
+				while (authCount < maxTries) {
+					authCount++
+					yield await authenticator.authenticate(resp as Response)
+				}
+
+				return false
+			}
+
+			// Attempt to authenticate until' generator has reached the max try threshold
+			for await (const didAuthenticate of tryAuthentication()) {
+				if (didAuthenticate) {
+					authSuccessful = true
+					break
+				}
+			}
+
+			// If we reached the max try threshold and authentication was still unsuccessful
+			// throw an exception
+			if (!authSuccessful) {
+				throw new Error('Authentication Failed')
+			}
+
+			// Replay request
+			resp = await finCsrfFetch(resource, options)
+		}
+	} else {
+		// If request doesn't contain an authenticator, continue has normal
+		resp = await finCsrfFetch(resource, options)
```

Comment:
The fetch is duplicated at both line 84 and 46 for no reason. If there's a fetch and it's successful then it shouldn't fetch again. 

I think you can just remove the else statement completely.